[Review, pending LazyLoad PR] Feature/typed mapTo

niekraaijmakers opened the pull request:
Typed MapTo for angular components.

Made backwards compatible by putting any as default generic type in editconfig.


For MapTo and LazyMapTo a generic can be passed:


```
MapTo<DownloadV1Model & MappedComponentProperties>('core-components-examples/wcm/angular/components/download')(DownloadV1Component, {isEmpty: DownloadV1IsEmptyFn});


const lazyEditConfig:EditConfig<LazyComponentType> = { isEmpty: (props) => !!props.otherValue };

LazyMapTo<LazyComponentType>('lazycomponent4')(() => import('../test/lazy-component-wrapper/lazy.component').then((m) => m.LazyComponent), lazyEditConfig:EditConfig);
```

Commit:
Make MapTo 1) only accept angular components 2) typed with a optional generic argument with gives you a typed EditConfig.

Commit:
Fixed copyright header

Commit:
Merge commit 'f85e2fac453151a345c6040d43cdb61aadbd0fce' into feature/typed-mapTo

* commit 'f85e2fac453151a345c6040d43cdb61aadbd0fce': (29 commits)
  chore(release): 1.0.7 [skip ci]
  fix(docs): remove --mode from docs build after a typedoc breaking change
  chore(deps): update dependency @adobe/aem-spa-page-model-manager to ~1.3.0
  style(releaserc): adapt code style to eslint warnings
  chore(release): 1.0.6 [skip ci]
  fix: docs generation command typo, releasing from dist directory
  chore(release): 1.0.5 [skip ci]
  fix: proper pkgRoot for semantic-release
  chore(release): 1.0.4 [skip ci]
  ci(tslint): add tslint config
  Revert "ci(actions): use eslint action"
  ci(actions): use eslint action
  fix(package.json): cleanup
  fix(eslint): adapt code according to eslint reports
  refactor(eslint): adjust code
  chore: cleanup
  refactor: cleanup
  chore(deps): update dependency @adobe/aem-spa-page-model-manager to ~1.2.0
  ci(npm): update token
  chore: update docs location and `.gitignore`
  ...

# Conflicts:
#	src/lib/layout/aem-component.directive.spec.ts
#	src/lib/layout/component-mapping.spec.ts
#	src/lib/layout/component-mapping.ts
#	src/lib/test/mapping.ts
#	src/lib/test/test-comp1.component.ts
#	src/lib/test/test-comp2.component.ts
#	src/lib/test/test-comp3.component.ts

Commit:
Linting fixes

Commit:
Merge branch 'feature/lazyloadable-components-2' into feature/typed-mapTo

* feature/lazyloadable-components-2:
  Linting fixes

# Conflicts:
#	src/lib/layout/aem-component.directive.ts
#	src/lib/layout/component-mapping.spec.ts
#	src/lib/layout/component-mapping.ts
#	src/lib/test/test-comp1.component.ts
#	src/lib/test/test-comp2.component.ts
#	src/lib/test/test-comp3.component.ts

Commit:
Remove linter warnings by adding typings

Commit:
feat: 🎸 Typed MapTo / LazyMapTo

MapTo and LazyMapTo are now typed with a generic extending
MappedComponentProperties. Use the interface MappedComponentProperties,
abstract class AbstractMappedComponent or the any type to fix your
build.

BREAKING CHANGE: 🧨 Generic Typed MapTo / LazyMapTo

Commit:
feat: 🎸 LazyMapTo support

Now you can lazy load your code splitted components on demand using a
Promise.

CezCz committed:
Update CHANGELOG.md

CezCz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+
+
+## Features
```
As per videocall, this changelog is generated automatically, no need to add it manually. Just put it in the git commit.

CezCz commented on package.json:
```diff
@@ -25,7 +25,7 @@
     "docs": "npm i && npx typedoc --excludePrivate ./src --out ./dist/docs --exclude '**/test/**/*' --exclude '**/*.spec.ts' --exclude src/test.ts",
     "e2e": "ng e2e",
-    "linter": "eslint .",
+    "linter": "eslint . --fix",
```
Please remove

CezCz commented on src/lib/layout/aem-component.directive.ts:
```diff
@@ -11,18 +11,23 @@
-  AfterViewInit,
-  OnInit,
-  OnDestroy, ChangeDetectorRef, OnChanges
+    AfterViewInit,
```
please restore original formatting

Commit:
Merge commit '0e6a09772169325637f07ea704f630ca371402e9' into feature/lazyloadable-components-2

* commit '0e6a09772169325637f07ea704f630ca371402e9':
  chore(deps): update dependency @adobe/aem-spa-component-mapping to ~1.1.0

Commit:
Merge commit '623be741d2243a86c06491081e9fc09a0d9dc7a5' into feature/lazyloadable-components-2

* commit '623be741d2243a86c06491081e9fc09a0d9dc7a5':
  Update CHANGELOG.md

Commit:
Adding default support for the lazyloading so imports can be simplified

Commit:
formatting

Commit:
formatting

Commit:
formatting

Commit:
formatting

Commit:
formatting

Commit:
`formatting

Commit:
Remove default approach, this does not work with AOT compilation (ivy) unfortunately.

Instead use this approach:

LazyMapTo('lazycomponent4')(() => import('../test/lazy-component-wrapper/lazy.component').then((m) => m.LazyComponent));

Commit:
Add / correct some comments

Commit:
Add unit test coverage for lazyloading directive

Commit:
Merge branch 'feature/lazyloadable-components-2' into feature/typed-mapTo

* feature/lazyloadable-components-2:
  Add unit test coverage for lazyloading directive
  Add / correct some comments
  Remove default approach, this does not work with AOT compilation (ivy) unfortunately.
  `formatting
  formatting
  formatting
  formatting
  formatting
  formatting
  Adding default support for the lazyloading so imports can be simplified
  chore(deps): update dependency @adobe/aem-spa-component-mapping to ~1.1.0
  Update CHANGELOG.md
  feat: 🎸 LazyMapTo support

# Conflicts:
#	src/lib/layout/aem-component.directive.spec.ts
#	src/lib/layout/aem-component.directive.ts
#	src/lib/layout/component-mapping.spec.ts
#	src/lib/layout/component-mapping.ts

Commit:
Add backwards compatibility by passing any as default generic

Commit:
Add isInEditor to component properties for compatibility with core components & ease of use
